actors/migration/top.go

```diff
@@ -129,60 +159,62 @@ func MigrateStateTree(ctx context.Context, store cbor.IpldStore, stateRootIn cid
 		return cid.Undef, err
 	}
 
-	// Extra setup
-	// miner
-	var p phoenix
-	if err := p.load(ctx, actorsIn); err != nil {
-		return cid.Undef, err
-	}
+	// Extra actor setup
 	// power
 	pm := migrations[builtin0.StoragePowerActorCodeID].StateMigration.(*powerMigrator)
 	pm.actorsIn = actorsIn
 
+	// Setup synchronization
+	errCh := make(chan error)
+	transferInCh := make(chan big.Int)
+	transferOutCh := make(chan big.Int)
+	transferCtx, transferCancel := context.WithCancel(ctx)
+	go func(ctx context.Context) {
+		// accumulate all burnt funds transfers
+		transferFromBurnt := big.Zero()
+		for {
+			select {
+			case transfer := <-transferInCh:
+				transferFromBurnt = big.Add(transferFromBurnt, transfer)
+			case <-ctx.Done():
+				transferOutCh <- transferFromBurnt
+			}
+		}
+
+	}(transferCtx)
+
 	// Iterate all actors in old state root
 	// Set new state root actors as we go
 	err = actorsIn.ForEach(func(addr address.Address, actorIn *states.Actor) error {
-		migration := migrations[actorIn.Code]
-
-		// This will be migrated at the end
-		if actorIn.Code == builtin0.VerifiedRegistryActorCodeID {
-			return nil
-		}
-		if actorIn.Code == builtin0.StorageMinerActorCodeID {
-			// setup migration fields
-			mm := migration.StateMigration.(*minerMigrator)
-			mm.MinerBalance = actorIn.Balance
-			mm.Transfer = big.Zero()
-		}
-		headOut, err := migration.StateMigration.MigrateState(ctx, store, actorIn.Head)
-		if err != nil {
-			return xerrors.Errorf("state migration error on %s actor at addr %s: %w", builtin.ActorNameByCode(migration.OutCodeCID), addr, err)
+		// Read from err channel without blocking and terminate on the first error
+		select {
+		case err := <-errCh:
+			return err
+		default:
 		}
 
-		// set up new state root with the migrated state
-		actorOut := states.Actor{
-			Code:       migration.OutCodeCID,
-			Head:       headOut,
-			CallSeqNum: actorIn.CallSeqNum,
-			Balance:    actorIn.Balance,
+		// Hand off migration of one actor, blocking if we are out of worker goroutines
+		if err := sem.Acquire(ctx, 1); err != nil {
+			return err
```

Comment:
Using a semaphore explicitly is a smell that we're not doing it right in Go.
- Consider that goroutines are _really_ cheap, try just not limiting them at all
- A bounded-capacity channel is probably a better pattern, with 1 producer that blocks when the channel is full